source/develop/release-management/features/sla/hosted-engine-agent-offloading.html.md

```diff
@@ -0,0 +1,53 @@
+---
+title: Hosted Engine Agent Offloading
+category: feature
+authors: dchaplyg
+feature_name: 'Hosted Engine: Move all the hardwork from agent to the broker'
+feature_status: In Development
+---
+
+# Hosted Engine: Move all the hardwork from agent to the broker
+
+## Summary
+
+Initially, when hosted engine tools was designed, they were intentionally split to the agent and the broker. Agent planned to be a lightweight application, responsible for cluster management decisions. And broker was supposed to make all the hard work - storage operations, monitoring, locking etc. 
+However, during hosted engine development, agent was burdened with additional load, like OVF extraction, storage preparation and other activities. 
+
+The goal of that feature is to make agent lightweight again and keep only management code in it, while moving all other stuff to the broker. This will result in a more maintenable code and more reliable behaviour of the hosted engine.
+## Owner
+
+*   Name: [Denis Chaplygin](https://github.com/akashihi)
+*   Email: <dchaplyg@redhat.com>
+
+## Detailed Description
+
+The specific goals associated with this feature are as follows:
+
+*   Move storage mounting/unmounting from agent to the broker.
+*   Broker should start/stop domain monitor on agent's command. Domain monitor could be implemented as submonitor.
+    
+    Domain monitoring needs to be changed drastically. When storage validation approach was changed (See [Documentation/External references](develop/release-management/features/sla/hosted-engine-agent-offloading/#documentation)), domain monitoring become an important for domain to be valied. Non-monitored domains are not reported as valid anymore. 
+    Therefore, old approach, when domains monitoring was active only outside of the LocalMaintenance mode is not applicable anymore. I propose it to be changed in the following way - domain monitoring still should be disabled in the LocalMaintenance mode. While domain monitoring is disabled, actual domain validation will be skipped and domain will be reported as valid. Broker will still try to update status (and read statues of the other cluster participants) even in the maintenance mode, so it will have a chance to block on broken storage. Thus, all I/O should be moved to a separate thread, so broker will not block on I/O issues, only that thread.
+*   Move sanlock acquire/release activities form agent to the broker.
+*   As steps above introduce new possible 'half-configured' state, FSM should be extended with that state.
+*   Agent should post state to the broker only and broker must publish it on storage if possible.
```

Comment:
- "Agent should post state to the broker only and broker must publish it on storage if possible." -> "The agent should only post state to the broker, and the broker must publish it on storage, if possible."

Reply:
Done